connectors/memory/memory.go

```diff
@@ -0,0 +1,493 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package memory
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"sort"
+	"sync"
+	"time"
+
+	"encoding/binary"
+
+	"github.com/satori/go.uuid"
+	"github.com/uber-go/dosa"
+	"github.com/uber-go/dosa/connectors/base"
+)
+
+// Connector is an in memory connector.
+// The in-memory connector stores it's data like this:
+// map[string]map[string][]map[string]dosa.FieldValue
+//
+// the first 'string' is the table name (entity name)
+// the second 'string' is the partition key, encoded using encoding/gob to guarantee uniqueness
+// within each 'partition' you have a list of rows ([]map[string]dosa.FieldValue)
+// these rows are kept ordered so that reads are lightning fast and searches are quick too
+// the row itself is a map of field name to value (map[string]dosaFieldValue])
+//
+// A read-write mutex lock is used to control concurrency, making reads work in parallel but
+// writes are not. There is no attempt to improve the concurrency of the read or write path by
+// adding more granular locks.
+type Connector struct {
+	base.Connector
+	data map[string]map[string][]map[string]dosa.FieldValue
+	lock sync.RWMutex
+}
+
+// partitionKeyBuilder extracts the partition key components from the map and encodes them,
+// generating a unique string. It uses the encoding/gob method to make a byte array as the
```

Comment:
I can't hash the value, because it can collide. This has to produce a unique string given a set of inputs